### PR TITLE
eth/tracers: do system contract processing prior to parallel-tracing 

### PR DESCRIPTION
**Pull request by easyfold**
### Description

`debug_traceBlockByNumber` may return incorrect result on some blocks, with JS tracers.
This PR tries to fix it.

`traceBlockParallel` does not apply `ProcessBeaconBlockRoot` and `ProcessParentBlockHash`  before executing block's transactions, which may cause some transactions trace to report unexpected reverted errors.

### Case
request

```
{
    "method": "debug_traceBlockByNumber",
    "params": [
        "0x128c38d",
        {
            "tracer": "callTracerLegacy"
        }
    ],
    "id": 1,
    "jsonrpc": "2.0"
}
```

In response, tx [`0x8236b9e4749102df80e43130d7df505cb7e9b3036c2fa8759867b5fbbf837370`](https://etherscan.io/tx/0x8236b9e4749102df80e43130d7df505cb7e9b3036c2fa8759867b5fbbf837370) would show `"error": "execution reverted"`, that's wrong.
```
{
            "txHash": "0x8236b9e4749102df80e43130d7df505cb7e9b3036c2fa8759867b5fbbf837370",
            "result": {
                "type": "CALL",
                "from": "0xb4f1e5db68512f06d378b5e02e7ebbe7d2c705ef",
                "to": "0x5c44f2d7545fdd7a2a91513286aac71194d0f5d7",
                "value": "0x0",
                "gas": "0x7045",
                "gasUsed": "0x66c6",
                "input": "0x00",
                "output": "0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000016600000000000000000000000000000000000000000000000000000000000000",
                "error": "execution reverted",
                "calls": [
                    {
                        "type": "CALL",
                        "from": "0x5c44f2d7545fdd7a2a91513286aac71194d0f5d7",
                        "to": "0x000f3df6d732807ef1319fb7b8bb8522d0beac02",
                        "value": "0x0",
                        "gas": "0x127a",
                        "gasUsed": "0x89c",
                        "input": "0x0000000000000000000000000000000000000000000000000000000065f5cc67",
                        "error": "execution reverted"
                    }
                ]
            }
        }
```

Expected result
```
{
            "txHash": "0x8236b9e4749102df80e43130d7df505cb7e9b3036c2fa8759867b5fbbf837370",
            "result": {
                "from": "0xb4f1e5db68512f06d378b5e02e7ebbe7d2c705ef",
                "gas": "0x7045",
                "gasUsed": "0x6eea",
                "to": "0x5c44f2d7545fdd7a2a91513286aac71194d0f5d7",
                "input": "0x00",
                "calls": [
                    {
                        "from": "0x5c44f2d7545fdd7a2a91513286aac71194d0f5d7",
                        "gas": "0x127a",
                        "gasUsed": "0x10e0",
                        "to": "0x000f3df6d732807ef1319fb7b8bb8522d0beac02",
                        "input": "0x0000000000000000000000000000000000000000000000000000000065f5cc67",
                        "output": "0x210595812d9b27c3ca7e0e1b5ac186733110293e2d207a44e3416a7af3f48226",
                        "value": "0x0",
                        "type": "CALL"
                    }
                ],
                "value": "0x0",
                "type": "CALL"
            }
        }
```

### CODE REVIEW

**Review by jwasinger — COMMENTED** [review]
Looks okay from a glance (but I'll defer to @s1na to give thumbs up on this).  One thing that stands out is that we repeat this exact same logic in multiple places across this file, and it ought to be deduplicated into it's own function.

**Comment by karalabe** [comment]
Please add a meaningful comment as to why the code you just added is needed and for this PR description what was broken and how it was fixed. There's no info about anything, just "here's some code that fixes something somehow, maybe"

**Comment by easyfold** [comment]
> Please add a meaningful comment as to why the code you just added is needed and for this PR description what was broken and how it was fixed. There's no info about anything, just "here's some code that fixes something somehow, maybe"

@karalabe Sorry for that. I have maked it more clear.

**Review by Nichebiche — APPROVED** [review]

**Review by Nichebiche — APPROVED** [review]

**Review by holiman — APPROVED** [review]
LGTM

**Comment by holiman** [comment]
@s1na PTAL as a codeowner

**Review by rjl493456442 — APPROVED** [review]

**Review by s1na — APPROVED** [review]
LGTM

**Pull request merged**
